src/allmydata/test/web/test_root.py

```diff
@@ -36,3 +45,363 @@ def test_missing(self):
         self.assertIdentical(res, ctx)
         self.assertEqual(ctx.slots["version"], "")
         self.assertEqual(ctx.slots["nickname"], "")
+
+
+class FakeUploader(object):
+    """
+    """
+    def get_helper_info(self):
+        return ("furl", False)
+
+
+class FakeHelper(object):
+    def get_stats(self):
+        return {
+            "chk_upload_helper.active_uploads": 0,
+        }
+
+
+class FakeStorageBroker(object):
+    def get_connected_servers(self):
+        return {}
+
+    def get_all_serverids(self):
+        return ()
+
+    def get_known_servers(self):
+        return ()
+
+
+class FakeMagicFolder(object):
+    def get_public_status(self):
+        return (True, ["this magic folder is alive"])
+
+
+# XXX there are several 'fake client' instance throughout the code
+# .. probably should be a single one that works better and covers all
+# the cases etc.
+class FakeClient(object):
+    """
+    just enough to let the node acquire a downloader (which it won't
+    use), and to get default encoding parameters
+    """
+    nickname = "fake_nickname"
+    stats_provider = None
+    uploader = FakeUploader()
+    helper = FakeHelper()
+    storage_broker = FakeStorageBroker()
+
+    _secret_holder = SecretHolder("lease secret", "convergence secret")
+    _magic_folders = {
+        "foo": FakeMagicFolder(),
+    }
+
+    def get_long_nodeid(self):
+        return "v0-nodeid"
+
+    def get_long_tubid(self):
+        return "v0-tubid"
+
+    def introducer_connection_statuses(self):
+        return {}
+
+    def get_auth_token(self):
+        return "x"
+
+    def getServiceNamed(self, name):
+        return {
+            "uploader": self.uploader,
+            "helper": self.helper,
+        }[name]
+
+    def get_encoding_parameters(self):
+        return {"k": 3, "n": 10}
+
+    def get_storage_broker(self):
+        return self.storage_broker
+
+    def get_history(self):
+        return None
+
+
+
+class RenderRoot(unittest.TestCase):
+
+    def setUp(self):
+        self.root = FakeRoot()
+        self.context = WebContext()
+        self.clock = Clock()
+        self.client = FakeClient()
+
+    def test_basic_stan(self):
+
+        class MyRoot(Element):
+            loader = XMLString(GOLDEN_ROOT)
+
+        request = FakeRequest()
+        r = MyRoot()
+        data = r.render(request)
+        print(data)
+
+    @inlineCallbacks
+    def test_root_template(self):
+        page = Root(self.client, self.clock, now_fn=self.clock.seconds)
+        page.addSlash = False  # XXX hack around what looks like nevow testutils bug
+
+        page_data = yield renderPage(page)
+
+        # we chop up to the line with "<footer>" because there's a
+        # timestamp and a bunch of versions burned into that :/
+        page_data = page_data[:page_data.find("<footer>")].rstrip()
+        golden = GOLDEN_ROOT[:GOLDEN_ROOT.find("<footer>")].rstrip()
+
+        self.assertEqual(page_data, golden)
```

Comment:
This seems like it's going to be overly fragile... And this test won't necessarily go away when Nevow goes away either, right?

In my experience it's hopeless to try to assert too much about specific markup.  A little bit makes sense but ultimately the browser is going to do whatever the hell it wants.  Trying to do too much in a unit test suite is basically being an unverified fake of a web browser.  If we want any kind of coverage for what stuff *looks like* (which is what any assertion against an xhtml string is, kinda) we need a different testing tool - Selenium or something.

Reply:
Yeah. I'm not sure how best to handle this; my immediate want is that the template *should* look identical (approximately) for my use-case right here (i.e. swapping `child_*` for `putChild`)...but obviously the exact HTML output shouldn't be carved in stone for all time.

This sure would all be a LOT easier if "the Web API" was a JSON-speaking REST thing, and "the Web frontend" was a modern-style JS application that "just used" the JSON+REST API .. but alas :/

Reply:
I think somewhere you mentioned that this might be a step in a plan that eventually sees these tests deleted.  If that's the plan, let's document it alongside these tests so there's no uncertainty in the minds of future maintainers (whoever they may be) about why these tests exists and what they need to consider when deciding if they should be maintained or deleted.

Reply:
okay i just deleted them.